drivers: Modified drivers to use DEVICE_AND_API_INIT()

varun-sha opened the pull request:
Modified drivers to use DEVICE_AND_API_INIT() instead of DEVICE_INIT()

This will make sure driver_api, is populated at build time and is exposed
to user space

Signed-off-by: Varun Sharma <varun.sharma@intel.com>

varun-sha committed:
drivers: Modify drivers to use DEVICE_AND_API_INIT()

Modified drivers to use DEVICE_AND_API_INIT() instead of DEVICE_INIT()

This will make sure driver_api,is populated at build time and is exposed
to user space

Signed-off-by: Varun Sharma <varun.sharma@intel.com>

vanwinkeljan reviewed (changes requested):
Looks like after this PR there is no use anymore for ```DEVICE_INIT``` as such remove ```DEVICE_INIT``` from ```device.h``` and update the documentation accordingly.

andrewboie commented:
> as such remove DEVICE_INIT from device.h and update the documentation accordingly.

This is a public API and will need to go through the deprecation process if we want to completely remove it.
You can tag it with `DEPRECATED_MACRO` and remove in two releases.

varun-sha committed:
samples: Modify samples to use DEVICE_AND_API_INIT()

tests: Modify tests to use DEVICE_AND_API_INIT()

Modified samples/tests drivers to use DEVICE_AND_API_INIT() instead
of deprecated DEVICE_INIT() api

Signed-off-by: Varun Sharma <varun.sharma@intel.com>

varun-sha commented:
> > as such remove DEVICE_INIT from device.h and update the documentation accordingly.
> 
> This is a public API and will need to go through the deprecation process if we want to completely remove it.
> You can tag it with `DEPRECATED_MACRO` and remove in two releases.

@andrewboie : DEVICE_INIT() is still being used in other places except drivers, like  in
SYS_INIT() defined in  include/init.h: SYS_INIT() and used at multiple places
&  in files: include/drivers/gpio/gpio_mmio32.h, boards/arm/nrf52_pca20020/board.c

Adding __DEPRECATED_MACRO triggers -werror & sanity check to fail, due to above usage's

So i have modified above usages to DEVICE_AND_API_INIT() also, please check.

andyross commented on boards/arm/nrf52_pca20020/board.c:
```diff
@@ -71,8 +72,8 @@ static const struct pwr_ctrl_cfg ccs_vdd_pwr_ctrl_cfg = {
+DEVICE_AND_API_INIT(ccs_vdd_pwr_ctrl_init, "", pwr_ctrl_init, NULL,
 	    &ccs_vdd_pwr_ctrl_cfg, POST_KERNEL,
-	    CONFIG_BOARD_CCS_VDD_PWR_CTRL_INIT_PRIORITY);
+	    CONFIG_BOARD_CCS_VDD_PWR_CTRL_INIT_PRIORITY, NULL);
```
Passing a NULL for api isn't doing anything but adding bytes to the source file.  This is just duplicating the behavior of DEVICE_INIT().

varun-sha replied:
@andyross  yes, plan is to deprecate DEVICE_INIT() so this change(using DEVICE_AND_API_INIT) is necessary, without this  error will returned as still using  DEVICE_INIT()  although deprecated.

Pls check: https://github.com/zephyrproject-rtos/zephyr/pull/11984#issuecomment-445937243

andyross commented on drivers/display/grove_lcd_rgb.c:
```diff
@@ -267,16 +267,6 @@ int glcd_initialize(struct device *port)
-	 * Since dev is probably still in registers.
-	 * use that to minimize code addition.
-	 */
-	port->driver_api = (void *)dev;
```
This driver was reusing its driver_data struct as the driver_api?   Is that kosher?  I know that this isn't related to this patch, but this seems like a bug to me.  Both are opaque handles, but the API struct is specific to the driver and driver_data should be one per device.

vanwinkeljan replied:
I also had a look at this and actually this driver has no API struct but you can not pass NULL else [device_get_binding](https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/device.c#L68) will not return the device even if it is available.

varun-sha replied:
@andyross : yeah since this driver necessitated setting of driver_api so that device_get_binding() passes, so driver owner just populated it with some magic number so grove_lcd can be referenced. Previously they were using (void *)dev, now i have used (void *)&grove_lcd_driver (struct glcd_data), which is available while using DEVICE_AND_API_INIT()

andyross commented on drivers/flash/spi_flash_w25qxxdv.c:
```diff
@@ -367,8 +367,8 @@ static int spi_flash_init(struct device *dev)
-	if (!ret) {
-		dev->driver_api = &spi_flash_api;
+	if (ret) {
+		return ret;
```
Probably pedantry, but this changes the logic: before, if the configure() call failed, the rest of the function would have been executed (which does some stuff related to "page layout").  Now that gets skipped.  That's probably OK, but given that this code that doesn't get tested under CI, it would be prudent to leave the logic unchanged for this kind of refactoring.  Or to make sure it gets tested (i.e. test the configure failure case!) on appropriate hardware, or at least get buy-in from someone responsible for the hardware.

varun-sha replied:
ok, so updated code, now returning in end only. this will not change current logic & if error is returned by spi_flash_wb_configure(),  init() function will return non zero value and driver_api will be set to NULL as per commit a68120de6d414117ddb67566a7397a5e23556841

andyross commented on tests/drivers/ipm/src/main.c:
```diff
@@ -39,9 +41,9 @@ static struct ipm_console_sender_config_info sender_config = {
+DEVICE_AND_API_INIT(ipm_console_send0, "ipm_send0", ipm_console_sender_init,
 	    NULL, &sender_config,
-	    APPLICATION, INIT_PRIO_IPM_SEND);
+	    APPLICATION, INIT_PRIO_IPM_SEND, NULL);
```
Needless "_INIT" and ", NULL"

varun-sha replied:
Can you pls elaborate?
modified DEVICE_INIT() to DEVICE_AND_API_INIT() as DEVICE_INIT() will tagged as deprecated

andyross commented on tests/drivers/ipm/src/main.c:
```diff
@@ -65,9 +67,9 @@ static struct ipm_console_receiver_config_info receiver_config = {
+DEVICE_AND_API_INIT(ipm_console_recv0, "ipm_recv0", ipm_console_receiver_init,
 	    &receiver_data, &receiver_config,
-	    APPLICATION, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+	    APPLICATION, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, NULL);
```
Same.  Why even bother having DEVICE_INIT() if you can't use it as intended?

varun-sha replied:
did you mean DEVICE_AND_API_INIT()  instead of DEVICE_INIT() in your comment? Can you pls elaborate?
I have modified DEVICE_INIT() to DEVICE_AND_API_INIT() as DEVICE_INIT() will tagged as deprecated, no driver_api is being used here, so assigned NULL

andyross reviewed (changes requested):
Very mild -1.  The idiom where you take a clear assignment to driver_api at initialization time and replace it with a macro argument makes perfect sense.  Where you get beyond that seems... less like a worthwhile change.

varun-sha commented:
> > as such remove DEVICE_INIT from device.h and update the documentation accordingly.
> 
> This is a public API and will need to go through the deprecation process if we want to completely remove it.
> You can tag it with `DEPRECATED_MACRO` and remove in two releases.

@vanwinkeljan Done, pls check

andyross commented:
OK, I get it. But... I dunno, my -1 remains mild, but if anything it's even stronger than before.

The only reason to deprecate DEIVCE_INIT() is because it does not set an API pointer and thus the resulting objects can't be discriminated by syscalls.   So old code that would have a NULL in that field now gets a warning.

Using the new style API but leaving a NULL in that field, as this patch does, now leaves the existing code  using the API incorrectly BUT NOT issuing a warning to get them fixed.  Either port the drivers to use a proper API struct or leave them in place with the deprecated API so that someone will find and fix it later.  All this does is hide the problem we're trying to fix.

varun-sha commented:
> OK, I get it. But... I dunno, my -1 remains mild, but if anything it's even stronger than before.
> 
> The only reason to deprecate DEIVCE_INIT() is because it does not set an API pointer and thus the resulting objects can't be discriminated by syscalls. So old code that would have a NULL in that field now gets a warning.
> Using the new style API but leaving a NULL in that field, as this patch does, now leaves the existing code using the API incorrectly BUT NOT issuing a warning to get them fixed. Either port the drivers to use a proper API struct or leave them in place with the deprecated API so that someone will find and fix it later. All this does is hide the problem we're trying to fix.

 @andyross : i understand your point, but if you see changes in boards/* & samples/*, tests/* , thes drivers are not having  driver_api as they don't need them,  so i don't want to tinker with them , which will be beyond scope of this and should be done by individual owners if required.
Additionally , if we  set __DEPRECATE_MACRO for DEVICE_INIT() &  keep both DEVICE_INIT & DEVICE_AND_API_INIT in code ,  our sanity uses -Wno-deprecated-declarations and returns error instead of warning if DEVICE_INIT() is used. 
So all usages should move to DEVICE_AND_API_INIT, if we want to deprecate DEVICE_INIT or we shouldn't add deprecate tag to DEVICE_INIT as of now.

API field is optional in DEVICE_AND_API_INIT() also & some drivers are  setting api as NULL, as they don't need driver_api as shared above, if we want to enforce driver_api to be set in DEVICE_AND_API_INIT()  then it should return error or warning if api is NULL.
As per current implementation  DEVICE_AND_API_INIT () does not give any warning if driver_api is NULL while device_get_binding.
https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/device.c#L85
@andyross Adding a warning message here, will resolve your concern?
@andrewboie @vanwinkeljan : your opinion...

andyross commented:
OK, I'm firming up here.  If our own source code is not ready to adhere to the requirements of the new API, **we should not be deprecating the old one yet**.

> API field is optional in DEVICE_AND_API_INIT()

Then why deprecate DEVICE_INIT() at all, if it's a legal and convenient shorthand for filling in a default value in an optional field?

vanwinkeljan commented:
@varun-sha No problem in keeping ```DEVICE_INIT```, initially I thought it was the intention to deprecate this macro.

varun-sha commented:
@andyross : address your review comment, please check

andyross reviewed (approved):
No more complaints.  Still seems like silly code motion to me on those NULL cases.
